# Writing Advanced SQL Queries

As part of this section we will understand how to write queries using some of the advanced features.

* Overview of Views
* Overview of Sub Queries
* CTAS - Create Table As Select
* Advanced DML Operations
* Merging or Upserting Data
* Pivoting Rows into Columns
* Overview of Analytic Functions
* Define Problem Statement – Top 5 Daily Products
* Analytic Functions – Aggregations
* Cumulative Aggregations
* Analytic Functions – Windowing
* Analytic Functions – Ranking
* Final Solution – Top 5 Daily Products

## Overview of Views
Here are the details related to views.
* View is nothing but a named query. We typically create views for most commonly used queries.
* Unlike tables, views does not physically store the data and when ever we write a query against view it will fetch the data from underlying tables defined as part of the views.
* We can perform DML operations over the tables via views with restrictions (for example, we cannot perform DML operations on views with joins, group by etc).
* Views that can be used to perform DML operations on underlying tables are called as **updatable views**
* Views can be used to provide restricted permissions on tables for DML Operations. However, it is not used these days.

In [ ]:
%load_ext sql

In [ ]:
%env DATABASE_URL=postgresql://itversity_retail_user:retail_password@localhost:5432/itversity_retail_db

In [ ]:
%%sql

CREATE OR REPLACE VIEW orders_v
AS
SELECT * FROM orders

In [ ]:
%%sql

CREATE VIEW orders_v
AS
SELECT * FROM orders

In [ ]:
%%sql

SELECT * FROM information_schema.tables
WHERE table_name ~ 'orders'

In [ ]:
%%sql

UPDATE orders_v
SET order_status = lower(order_status)

In [ ]:
%%sql

SELECT * FROM orders LIMIT 10

In [ ]:
%%sql

UPDATE orders_v
SET order_status = upper(order_status)

In [ ]:
%%sql

CREATE OR REPLACE VIEW order_details_v
AS
SELECT * FROM orders o
    JOIN order_items oi
        ON o.order_id = oi.order_item_order_id

In [ ]:
%%sql

SELECT * FROM order_details_v LIMIT 10

In [ ]:
%%sql

SELECT count(1) FROM order_details_v

In [ ]:
%%sql

SELECT order_date,
    order_item_product_id,
    round(sum(order_item_subtotal)::numeric, 2) AS revenue
FROM order_details_v 
GROUP BY order_date,
    order_item_product_id
ORDER BY order_date,
    revenue DESC
LIMIT 10

In [ ]:
%%sql

SELECT * FROM order_details_v
WHERE order_id = 2

```{note}
We cannot directly update data in tables via views when the view is defined with joins. Even operations such as `GROUP BY` or `ORDER BY` etc will make views not updatable by default.
```

In [ ]:
%%sql

UPDATE order_details_v
SET
    order_status = 'pending_payment'
WHERE order_id = 2

## Named Queries - Using WITH Clause

Let us understand how to use `WITH` clause to define a named query.
* At times we might have to develop a large query in which same complex logic need to be used multiple times. The query can become cumbersome if you just define the same logic multiple times.
* One of the way to mitigate that issue is by providing the name to the logic using WITH clause.
* We can only use the names provided to named queries as part of the main query which follows the WITH clause.

```{note}
In case of frequently used complex and large query, we use named queries while defining the views. We will then use view for reporting purposes.
```

In [ ]:
%load_ext sql

In [ ]:
%env DATABASE_URL=postgresql://itversity_retail_user:retail_password@localhost:5432/itversity_retail_db

In [ ]:
%%sql

WITH order_details_nq AS (
    SELECT * FROM orders o
        JOIN order_items oi
            on o.order_id = oi.order_item_order_id
) SELECT * FROM order_details_nq LIMIT 10

```{error}
One cannot use the named queries apart from the query in which it is defined. Following query will fail.
```

In [ ]:
%%sql

SELECT * FROM order_details_nq LIMIT 10

In [ ]:
%%sql

WITH order_details_nq AS (
    SELECT * FROM orders o
        JOIN order_items oi
            on o.order_id = oi.order_item_order_id
) SELECT order_date,
    order_item_product_id,
    round(sum(order_item_subtotal)::numeric, 2) AS revenue
FROM order_details_nq 
GROUP BY order_date,
    order_item_product_id
ORDER BY order_date,
    revenue DESC
LIMIT 10

In [ ]:
%%sql

CREATE OR REPLACE VIEW daily_product_revenue_v
AS
WITH order_details_nq AS (
    SELECT * FROM orders o
        JOIN order_items oi
            on o.order_id = oi.order_item_order_id
) SELECT order_date,
    order_item_product_id,
    round(sum(order_item_subtotal)::numeric, 2) AS revenue
FROM order_details_nq 
GROUP BY order_date,
    order_item_product_id

In [ ]:
%%sql

SELECT * FROM daily_product_revenue_v
ORDER BY order_date, revenue DESC
LIMIT 10

## Overview of Sub Queries
Let us understand details related to Sub Queries. We will also briefly discuss about nested sub queries.

* We can have queries in from clause and such queries are called as sub queries.
* Sub queries are commonly used with queries using analytic functions to filter the data further. We will see details after going through analytic functions as part of this section.
* It is mandatory to have alias for the sub query.
* Sub queries can also be used in `WHERE` clause with `IN` as well as `EXISTS`. As part of the sub query we can have join like conditions between tables in `FROM` clause of the main query and sub query. Such queries are called as **Nested Sub Queries**.

In [ ]:
%load_ext sql

In [ ]:
%env DATABASE_URL=postgresql://itversity_retail_user:retail_password@localhost:5432/itversity_retail_db

```{note}
Simplest example for a subquery
```

In [ ]:
%%sql

SELECT * FROM (SELECT current_date) AS q

```{note}
Realistic example for a subquery. We will get into details related to this query after covering analytic functions
```

In [ ]:
%%sql

SELECT * FROM (
    SELECT nq.*,
        dense_rank() OVER (
            PARTITION BY order_date
            ORDER BY revenue DESC
        ) AS drnk
    FROM (
        SELECT o.order_date,
            oi.order_item_product_id,
            round(sum(oi.order_item_subtotal)::numeric, 2) AS revenue
        FROM orders o 
            JOIN order_items oi
                ON o.order_id = oi.order_item_order_id
        WHERE o.order_status IN ('COMPLETE', 'CLOSED')
        GROUP BY o.order_date, oi.order_item_product_id
    ) nq
) nq1
WHERE drnk <= 5
ORDER BY order_date, revenue DESC
LIMIT 20

```{note}
Multiple realistic examples for nested sub queries. You can see example with `IN` as well as `EXISTS` operators.
```

In [ ]:
%%sql

SELECT * FROM order_items oi
WHERE oi.order_item_order_id 
    NOT IN (
        SELECT order_id FROM orders o
        WHERE o.order_id = oi.order_item_order_id
    )
LIMIT 10

In [ ]:
%%sql

SELECT count(1) FROM order_items oi
WHERE oi.order_item_order_id 
    IN (
        SELECT order_id FROM orders o
        WHERE o.order_id = oi.order_item_order_id
    )
LIMIT 10

In [ ]:
%%sql

SELECT * FROM order_items oi
WHERE NOT EXISTS (
        SELECT 1 FROM orders o
        WHERE o.order_id = oi.order_item_order_id
    )
LIMIT 10

In [ ]:
%%sql

SELECT * FROM order_items oi
WHERE EXISTS (
        SELECT 1 FROM orders o
        WHERE o.order_id = oi.order_item_order_id
    )
LIMIT 10

## CTAS - Create Table as Select

Let us understand details related to CTAS or Create Table As Select.

* CTAS is primarily used to create tables based on query results.
* Following are some of the use cases for which we typically use CTAS.
  * Taking back up of tables for troubleshooting and debugging performance issues.
  * Reorganizing the tables for performance tuning.
  * Getting query results into a table for data analysis as well as checking data quality.
* We cannot specify column names and data types as part of `CREATE TABLE` clause in CTAS. It will pick the column names from the `SELECT` clause.
* It is a good practice to specify meaningful aliases as part of the `SELECT` clause for derived values.
* Also it is a good practice to explicitly type cast to the desired data type for derived values.

In [ ]:
%load_ext sql

In [ ]:
%env DATABASE_URL=postgresql://itversity_retail_user:retail_password@localhost:5432/itversity_retail_db

In [ ]:
%%sql

DROP TABLE IF EXISTS customers_backup

In [ ]:
%%sql

CREATE TABLE customers_backup
AS
SELECT * FROM customers

In [ ]:
%%sql

DROP TABLE IF EXISTS orders_backup

In [ ]:
%%sql

CREATE TABLE orders_backup
AS
SELECT order_id,
    to_char(order_date, 'yyyy')::int AS order_year,
    to_char(order_date, 'MM')::int AS order_month,
    to_char(order_date, 'dd')::int AS order_day_of_month,
    to_char(order_date, 'DDD')::int AS order_day_of_year,
    order_customer_id,
    order_status
FROM orders

In [ ]:
%%sql

SELECT * FROM orders_backup LIMIT 10

```{note}
At times we have to create empty table with only structure of the table. We can specify always false condition such as `1 = 2` as part of `WHERE` clause using CTAS.
```

In [ ]:
%%sql

DROP TABLE IF EXISTS order_items_empty

In [ ]:
%%sql

CREATE TABLE order_items_empty
AS
SELECT * FROM order_items WHERE 1 = 2

In [ ]:
%%sql

SELECT count(1) FROM order_items_empty

```{note}
Keeping databases clean is very important. It is a good practice to clean up any temporary tables created for learning or troubleshooting issues.

In this case all the tables created using CTAS are dropped
```

In [ ]:
%%sql

DROP TABLE IF EXISTS customers_backup;
DROP TABLE IF EXISTS orders_backup;
DROP TABLE IF EXISTS order_items_empty;

## Advanced DML Operations

As we gain enough knowledge related to writing queries, let us explore some advanced DML Operations.
* We can insert query results into a table using `INSERT` with `SELECT`.
* As long as columns specified for table in `INSERT` statement and columns projected in `SELECT` clause match, it works.
* We can also use query results for `UPDATE` as well as `DELETE`.

In [ ]:
%load_ext sql

In [ ]:
%env DATABASE_URL=postgresql://itversity_retail_user:retail_password@localhost:5432/itversity_retail_db

```{note}
Creating customer order metrics table to demonstrate advanced DML Operations. We will also add primary key to this table. We will be storing number of orders placed and revenue generated for each customer in a given month.
```

In [ ]:
%%sql

DROP TABLE IF EXISTS customer_order_metrics_mthly

In [ ]:
%%sql

CREATE TABLE customer_order_metrics_mthly (
    customer_id INT,
    order_month CHAR(7),
    order_count INT,
    order_revenue FLOAT
)

In [ ]:
%%sql

ALTER TABLE customer_order_metrics_mthly
    ADD PRIMARY KEY (order_month, customer_id)

```{note}
Here is the query to get monthly customer orders metrics. First we will be inserting customer_id, order_month and order_count into the table. 
```

```{warning}
If the below query is run multiple times, every time data in both orders and order_items need to be processed. As the data volumes grow the query uses considerable amount of resources. It will be better if we can pre-aggregate the data.
```

In [ ]:
%%sql

SELECT o.order_customer_id,
    to_char(o.order_date, 'yyyy-MM') AS order_month,
    count(1) AS order_count,
    round(sum(order_item_subtotal)::numeric, 2) AS order_revenue
FROM orders o 
    JOIN order_items oi
        ON o.order_id = oi.order_item_order_id
GROUP BY o.order_customer_id,
    to_char(o.order_date, 'yyyy-MM')
ORDER BY order_month,
    order_count DESC
LIMIT 10

```{warning}
Here are the number of records that need to be processed every time. Also it involves expensive join.
```

In [ ]:
%%sql

SELECT count(1)
FROM orders o 
    JOIN order_items oi
        ON o.order_id = oi.order_item_order_id

```{note}
Let us first insert the data into the table with out revenue. We will update the revenue later as an example for updating using query results.
```

In [ ]:
%%sql

INSERT INTO customer_order_metrics_mthly
SELECT o.order_customer_id,
    to_char(o.order_date, 'yyyy-MM') AS order_month,
    count(1) order_count,
    NULL
FROM orders o 
    JOIN order_items oi
        ON o.order_id = oi.order_item_order_id
GROUP BY o.order_customer_id,
    to_char(o.order_date, 'yyyy-MM')

In [ ]:
%%sql

SELECT * FROM customer_order_metrics_mthly
ORDER BY order_month,
    customer_id
LIMIT 10

```{note}
Updating order_revenue along with count. This is expensive operation, but we will be running only once.
```

In [ ]:
%%sql

UPDATE customer_order_metrics_mthly comd
SET 
    (order_count, order_revenue) = (
        SELECT count(1),
            round(sum(order_item_subtotal)::numeric, 2)
        FROM orders o 
            JOIN order_items oi
                ON o.order_id = oi.order_item_order_id
        WHERE o.order_customer_id = comd.customer_id
            AND to_char(o.order_date, 'yyyy-MM') = comd.order_month
            AND to_char(o.order_date, 'yyyy-MM') = '2013-08'
            AND comd.order_month = '2013-08'
        GROUP BY o.order_customer_id,
            to_char(o.order_date, 'yyyy-MM')
    )
WHERE EXISTS (
    SELECT 1 FROM orders o
    WHERE o.order_customer_id = comd.customer_id
        AND to_char(o.order_date, 'yyyy-MM') = comd.order_month
        AND to_char(o.order_date, 'yyyy-MM') = '2013-08'
) AND comd.order_month = '2013-08'

```{note}
As data is pre processed and loaded into the table, queries similar to below ones against **customer_order_metrics_mthly** will run much faster.

We need to process lesser amount of data with out expensive join.
```

In [ ]:
%%sql

SELECT * FROM customer_order_metrics_mthly
WHERE order_month = '2013-08'
ORDER BY order_month,
    customer_id
LIMIT 10

```{note}
As an example for delete using query, we will delete all the dormant customers from **customers** table. Dormant customers are those customers who never placed any order. For this we will create back up customers table as I do not want to play with customers.
```

In [ ]:
%%sql

DROP TABLE IF EXISTS customers_backup

In [ ]:
%%sql

CREATE TABLE customers_backup
AS
SELECT * FROM customers

In [ ]:
%%sql

SELECT count(1) FROM customers_backup c
    LEFT OUTER JOIN orders o
        ON c.customer_id = o.order_customer_id
WHERE o.order_customer_id IS NULL

In [ ]:
%%sql

SELECT count(1) FROM customers_backup c
WHERE NOT EXISTS (
    SELECT 1 FROM orders o
    WHERE c.customer_id = o.order_customer_id
)

```{note}
We need to use nested sub queries as part of the delete with `NOT EXISTS` or `NOT IN` as demonstrated below. We cannot use direct joins as part of the `DELETE`.
```

In [ ]:
%%sql

DELETE FROM customers_backup c
WHERE NOT EXISTS (
    SELECT 1 FROM orders o
    WHERE c.customer_id = o.order_customer_id
)

In [ ]:
%%sql

SELECT count(1) FROM customers_backup

In [ ]:
%%sql

DELETE FROM customers_backup c
WHERE customer_id NOT IN (
    SELECT order_customer_id FROM orders o
)

## Merging or Upserting Data

At times we need to merge or upsert the data (update existing records and insert new records)
* One of the way to achieve merge or upsert is to develop 2 statements - one to update and other to insert.
* The queries in both the statements (update and insert) should return mutually exclusive results. 
* Even though the statements can be executed in any order, updating first and then inserting perform better in most of the cases (as update have to deal with lesser number of records with this approach)
* We can also take care of merge or upsert using `INSERT` with `ON CONFLICT (columns) DO UPDATE`.
* Postgres does not have either `MERGE` or `UPSERT` as part of the SQL syntax.

In [ ]:
%load_ext sql

In [ ]:
%env DATABASE_URL=postgresql://itversity_retail_user:retail_password@localhost:5432/itversity_retail_db

In [ ]:
%sql DROP TABLE IF EXISTS customer_order_metrics_dly

In [ ]:
%%sql

CREATE TABLE customer_order_metrics_dly (
    customer_id INT,
    order_date DATE,
    order_count INT,
    order_revenue FLOAT
)

In [ ]:
%%sql

ALTER TABLE customer_order_metrics_dly
    ADD PRIMARY KEY (customer_id, order_date)

```{note}
Let us go through the 2 statement approach. Here we are inserting data for the month of August 2013.
```

In [ ]:
%%sql

INSERT INTO customer_order_metrics_dly
SELECT o.order_customer_id,
    o.order_date,
    count(1) order_count,
    NULL
FROM orders o 
    JOIN order_items oi
        ON o.order_id = oi.order_item_order_id
WHERE o.order_date BETWEEN '2013-08-01' AND '2013-08-31'
GROUP BY o.order_customer_id,
    o.order_date

```{note}
Now we want to merge data into the table using 2013 August to 2013 October. As we are using 2 statement approach, first we should update and then we should insert
```

In [ ]:
%%sql

UPDATE customer_order_metrics_dly comd
SET 
    (order_count, order_revenue) = (
        SELECT count(1),
            round(sum(oi.order_item_subtotal)::numeric, 2)
        FROM orders o 
            JOIN order_items oi
                ON o.order_id = oi.order_item_order_id
        WHERE o.order_date BETWEEN '2013-08-01' AND '2013-10-31'
            AND o.order_customer_id = comd.customer_id
            AND o.order_date = comd.order_date
        GROUP BY o.order_customer_id,
            o.order_date
    )
WHERE comd.order_date BETWEEN '2013-08-01' AND '2013-10-31'

In [ ]:
%%sql

SELECT * FROM customer_order_metrics_dly
ORDER BY order_date, customer_id
LIMIT 10

In [ ]:
%%sql

SELECT to_char(order_date, 'yyyy-MM'), count(1) FROM customer_order_metrics_dly
GROUP BY to_char(order_date, 'yyyy-MM')
LIMIT 10

In [ ]:
%%sql

INSERT INTO customer_order_metrics_dly
SELECT o.order_customer_id AS customer_id,
    o.order_date,
    count(1) order_count,
    round(sum(order_item_subtotal)::numeric, 2)
FROM orders o 
    JOIN order_items oi
        ON o.order_id = oi.order_item_order_id
WHERE o.order_date BETWEEN '2013-08-01' AND '2013-10-31'
    AND NOT EXISTS (
        SELECT 1 FROM customer_order_metrics_dly codm
        WHERE o.order_customer_id = codm.customer_id
            AND o.order_date = codm.order_date
    )
GROUP BY o.order_customer_id,
    o.order_date

In [ ]:
%%sql

SELECT * FROM customer_order_metrics_dly
WHERE order_date::varchar ~ '2013-09'
ORDER BY order_date, customer_id
LIMIT 10

In [ ]:
%%sql

SELECT to_char(order_date, 'yyyy-MM'), count(1) FROM customer_order_metrics_dly
GROUP BY to_char(order_date, 'yyyy-MM')
LIMIT 10

```{note}
Let us see how we can upsert or merge the data using `INSERT` with `ON CONFLICT (columns) DO UPDATE`. We will first insert data for the month of August 2013 and then upsert or merge for the months of August 2013 to October 2013.
```

In [ ]:
%sql TRUNCATE TABLE customer_order_metrics_dly

In [ ]:
%%sql

INSERT INTO customer_order_metrics_dly
SELECT o.order_customer_id,
    o.order_date,
    count(1) order_count,
    NULL
FROM orders o 
    JOIN order_items oi
        ON o.order_id = oi.order_item_order_id
WHERE o.order_date BETWEEN '2013-08-01' AND '2013-08-31'
GROUP BY o.order_customer_id,
    o.order_date

```{note}
We need to have unique or primary key constraint on the columns specified as part of `ON CONFLICT` clause.
```

In [ ]:
%%sql

ALTER TABLE customer_order_metrics_dly DROP CONSTRAINT customer_order_metrics_dly_pkey

In [ ]:
%%sql

ALTER TABLE customer_order_metrics_dly
    ADD PRIMARY KEY (customer_id, order_date)

In [ ]:
%%sql

INSERT INTO customer_order_metrics_dly
SELECT o.order_customer_id,
    o.order_date,
    count(1) order_count,
    round(sum(order_item_subtotal)::numeric, 2) AS order_revenue
FROM orders o 
    JOIN order_items oi
        ON o.order_id = oi.order_item_order_id
WHERE o.order_date BETWEEN '2013-08-01' AND '2013-10-31'
GROUP BY o.order_customer_id,
    o.order_date
ON CONFLICT (customer_id, order_date) DO UPDATE SET
    order_count = EXCLUDED.order_count,
    order_revenue = EXCLUDED.order_revenue

In [ ]:
%%sql

SELECT * FROM customer_order_metrics_dly
WHERE order_date::varchar ~ '2013-09'
ORDER BY order_date, customer_id
LIMIT 10

In [ ]:
%%sql

SELECT to_char(order_date, 'yyyy-MM'), count(1) FROM customer_order_metrics_dly
GROUP BY to_char(order_date, 'yyyy-MM')
LIMIT 10

## Pivoting Rows into Columns

Let us understand how we can pivot rows into columns in Postgres.
* Actual results

|order_date|order_status|count|
|----------|------------|-----|
|2013-07-25 00:00:00|CANCELED|1|
|2013-07-25 00:00:00|CLOSED|20|
|2013-07-25 00:00:00|COMPLETE|42|
|2013-07-25 00:00:00|ON_HOLD|5|
|2013-07-25 00:00:00|PAYMENT_REVIEW|3|
|2013-07-25 00:00:00|PENDING|13|
|2013-07-25 00:00:00|PENDING_PAYMENT|41|
|2013-07-25 00:00:00|PROCESSING|16|
|2013-07-25 00:00:00|SUSPECTED_FRAUD|2|
|2013-07-26 00:00:00|CANCELED|3|
|2013-07-26 00:00:00|CLOSED|29|
|2013-07-26 00:00:00|COMPLETE|87|
|2013-07-26 00:00:00|ON_HOLD|19|
|2013-07-26 00:00:00|PAYMENT_REVIEW|6|
|2013-07-26 00:00:00|PENDING|31|
|2013-07-26 00:00:00|PENDING_PAYMENT|59|
|2013-07-26 00:00:00|PROCESSING|30|
|2013-07-26 00:00:00|SUSPECTED_FRAUD|5|

* Pivoted results

|order_date|CANCELED|CLOSED|COMPLETE|ON_HOLD|PAYMENT_REVIEW|PENDING|PENDING_PAYMENT|PROCESSING|SUSPECTED_FRAUD|
|----------|--------|------|--------|-------|--------------|-------|---------------|----------|---------------|
|2013-07-25|1|20|42|5|3|13|41|16|2|
|2013-07-26|3|29|87|19|6|31|59|30|5|

* We need to use `crosstab` as part of `FROM` clause to pivot the data. We need to pass the main query to `crosstab` function.
* We need to install `tablefunc` as Postgres superuser to expose functions like crosstab - `CREATE EXTENSION tablefunc;`

```{note}
If you are using environment provided by us, you don't need to install `tablefunc`. If you are using your own environment run this command by logging in as superuser into postgres server to install `tablefunc`.

`CREATE EXTENSION tablefunc;`

However, in some cases you might have to run scripts in postgres. Follow official instructions by searching around.
```

In [ ]:
%load_ext sql

In [ ]:
%env DATABASE_URL=postgresql://itversity_retail_user:retail_password@localhost:5432/itversity_retail_db

In [ ]:
%%sql

SELECT order_date,
    order_status,
    count(1)
FROM orders
GROUP BY order_date,
    order_status
ORDER BY order_date,
    order_status
LIMIT 18

In [ ]:
%%sql

SELECT * FROM crosstab(
    'SELECT order_date,
        order_status,
        count(1) AS order_count
    FROM orders
    GROUP BY order_date,
        order_status',
    'SELECT DISTINCT order_status FROM orders ORDER BY 1'
) AS (
    order_date DATE,
    "CANCELED" INT,
    "CLOSED" INT,
    "COMPLETE" INT,
    "ON_HOLD" INT,
    "PAYMENT_REVIEW" INT,
    "PENDING" INT,
    "PENDING_PAYMENT" INT,
    "PROCESSING" INT,
    "SUSPECTED_FRAUD" INT
)
LIMIT 10

## Overview of Analytic Functions

Let us get an overview of Analytics or Windowing Functions as part of **SQL**.

* Aggregate Functions (`sum`, `min`, `max`, `avg`)
* Window Functions (`lead`, `lag`, `first_value`, `last_value`)
* Rank Functions (`rank`, `dense_rank`, `row_number` etc)
* For all the functions when used as part of Analytic or Windowing functions we use `OVER` clause.
* For aggregate functions we typically use `PARTITION BY`
* For global ranking and windowing functions we can use `ORDER BY sort_column` and for ranking and windowing with in a partition or group we can use `PARTITION BY partition_column ORDER BY sort_column`.
* Here is how the syntax will look like.
  * Aggregate - `func() OVER (PARTITION BY partition_column)`
  * Global Rank - `func() OVER (ORDER BY sort_column DESC)`
  * Rank in a partition - `func() OVER (PARTITION BY partition_column ORDER BY sort_column DESC)`
* We can also get cumulative or moving metrics by adding `ROWS BETWEEN` clause. We will see details later.

### Prepare Tables

Let us create couple of tables which will be used for the demonstrations of Windowing and Ranking functions.

* We have **ORDERS** and **ORDER_ITEMS** tables in our retail database.
* Let us take care of computing daily revenue as well as daily product revenue.
* As we will be using same data set several times, let us create the tables to pre compute the data.
* **daily_revenue** will have the **order_date** and **revenue**, where data is aggregated using **order_date** as partition key.
* **daily_product_revenue** will have **order_date**, **order_item_product_id** and **revenue**. In this case data is aggregated using **order_date** and **order_item_product_id** as partition keys.

```{note}
Let us create table using CTAS to save daily revenue.
```

In [1]:
%load_ext sql

In [2]:
%env DATABASE_URL=postgresql://itversity_retail_user:retail_password@localhost:5432/itversity_retail_db

env: DATABASE_URL=postgresql://itversity_retail_user:retail_password@localhost:5432/itversity_retail_db


In [4]:
%%sql

DROP TABLE IF EXISTS daily_revenue

 * postgresql://itversity_retail_user:***@localhost:5432/itversity_retail_db
Done.


[]

In [5]:
%%sql

CREATE TABLE daily_revenue
AS
SELECT o.order_date,
    round(sum(oi.order_item_subtotal)::numeric, 2) AS revenue
FROM orders o JOIN order_items oi
ON o.order_id = oi.order_item_order_id
WHERE o.order_status IN ('COMPLETE', 'CLOSED')
GROUP BY o.order_date

 * postgresql://itversity_retail_user:***@localhost:5432/itversity_retail_db
364 rows affected.


[]

In [7]:
%%sql

SELECT * FROM daily_revenue
ORDER BY order_date
LIMIT 10

 * postgresql://itversity_retail_user:***@localhost:5432/itversity_retail_db
10 rows affected.


[(datetime.datetime(2013, 7, 25, 0, 0), Decimal('31547.23')),
 (datetime.datetime(2013, 7, 26, 0, 0), Decimal('54713.23')),
 (datetime.datetime(2013, 7, 27, 0, 0), Decimal('48411.48')),
 (datetime.datetime(2013, 7, 28, 0, 0), Decimal('35672.03')),
 (datetime.datetime(2013, 7, 29, 0, 0), Decimal('54579.70')),
 (datetime.datetime(2013, 7, 30, 0, 0), Decimal('49329.29')),
 (datetime.datetime(2013, 7, 31, 0, 0), Decimal('59212.49')),
 (datetime.datetime(2013, 8, 1, 0, 0), Decimal('49160.08')),
 (datetime.datetime(2013, 8, 2, 0, 0), Decimal('50688.58')),
 (datetime.datetime(2013, 8, 3, 0, 0), Decimal('43416.74'))]

```{note}
Let us create table using CTAS to save daily product revenue.
```

In [8]:
%%sql

DROP TABLE IF EXISTS daily_product_revenue

 * postgresql://itversity_retail_user:***@localhost:5432/itversity_retail_db
Done.


[]

In [9]:
%%sql

CREATE TABLE daily_product_revenue
AS
SELECT o.order_date,
    oi.order_item_product_id,
    round(sum(oi.order_item_subtotal)::numeric, 2) AS revenue
FROM orders o JOIN order_items oi
ON o.order_id = oi.order_item_order_id
WHERE o.order_status IN ('COMPLETE', 'CLOSED')
GROUP BY o.order_date, oi.order_item_product_id

 * postgresql://itversity_retail_user:***@localhost:5432/itversity_retail_db
9120 rows affected.


[]

In [10]:
%%sql

SELECT * FROM daily_product_revenue
ORDER BY order_date, revenue DESC
LIMIT 10

 * postgresql://itversity_retail_user:***@localhost:5432/itversity_retail_db
10 rows affected.


[(datetime.datetime(2013, 7, 25, 0, 0), 1004, Decimal('5599.72')),
 (datetime.datetime(2013, 7, 25, 0, 0), 191, Decimal('5099.49')),
 (datetime.datetime(2013, 7, 25, 0, 0), 957, Decimal('4499.70')),
 (datetime.datetime(2013, 7, 25, 0, 0), 365, Decimal('3359.44')),
 (datetime.datetime(2013, 7, 25, 0, 0), 1073, Decimal('2999.85')),
 (datetime.datetime(2013, 7, 25, 0, 0), 1014, Decimal('2798.88')),
 (datetime.datetime(2013, 7, 25, 0, 0), 403, Decimal('1949.85')),
 (datetime.datetime(2013, 7, 25, 0, 0), 502, Decimal('1650.00')),
 (datetime.datetime(2013, 7, 25, 0, 0), 627, Decimal('1079.73')),
 (datetime.datetime(2013, 7, 25, 0, 0), 226, Decimal('599.99'))]

## Define Problem Statement – Top 5 Daily Products

Let us define problem statement for which we will try to come up with solution using powerful analytic functions.

## Analytic Functions – Aggregations

Let us see how we can perform aggregations with in a partition or group using Windowing/Analytics Functions.

* For simple aggregations where we have to get grouping key and aggregated results we can use **GROUP BY**.
* If we want to get the raw data along with aggregated results, then using **GROUP BY** is not possible or overly complicated.
* Using aggregate functions with **OVER** Clause not only simplifies the process of writing query, but also better with respect to performance.
* Let us take an example of getting employee salary percentage when compared to department salary expense.

```{warning}
If you are using Jupyter based environment make sure to restart the kernel, as the session might have been already connected with retail database.
```

In [1]:
%load_ext sql

In [2]:
%env DATABASE_URL=postgresql://itversity_hr_user:hr_password@localhost:5432/itversity_hr_db

env: DATABASE_URL=postgresql://itversity_hr_user:hr_password@localhost:5432/itversity_hr_db


In [3]:
%%sql

SELECT employee_id, department_id, salary 
FROM employees 
ORDER BY department_id, salary
LIMIT 10

10 rows affected.


[(200, 10, Decimal('4400.00')),
 (202, 20, Decimal('6000.00')),
 (201, 20, Decimal('13000.00')),
 (119, 30, Decimal('2500.00')),
 (118, 30, Decimal('2600.00')),
 (117, 30, Decimal('2800.00')),
 (116, 30, Decimal('2900.00')),
 (115, 30, Decimal('3100.00')),
 (114, 30, Decimal('11000.00')),
 (203, 40, Decimal('6500.00'))]

```{note}
Let us write the query using `GROUP BY` approach.
```

In [4]:
%%sql

SELECT department_id,
    sum(salary) AS department_salary_expense
FROM employees
GROUP BY department_id
ORDER BY department_id

 * postgresql://itversity_hr_user:***@localhost:5432/itversity_hr_db
12 rows affected.


[(10, Decimal('4400.00')),
 (20, Decimal('19000.00')),
 (30, Decimal('24900.00')),
 (40, Decimal('6500.00')),
 (50, Decimal('156400.00')),
 (60, Decimal('28800.00')),
 (70, Decimal('10000.00')),
 (80, Decimal('304500.00')),
 (90, Decimal('58000.00')),
 (100, Decimal('51600.00')),
 (110, Decimal('20300.00')),
 (None, Decimal('7000.00'))]

In [7]:
%%sql

SELECT e.employee_id, e.department_id, e.salary,
    ae.department_salary_expense,
    ae.avg_salary_expense
FROM employees e JOIN (
    SELECT department_id, 
        sum(salary) AS department_salary_expense,
        round(avg(salary)::numeric, 2) AS avg_salary_expense
    FROM employees
    GROUP BY department_id
) ae
ON e.department_id = ae.department_id
ORDER BY department_id, salary
LIMIT 10

 * postgresql://itversity_hr_user:***@localhost:5432/itversity_hr_db
10 rows affected.


[(200, 10, Decimal('4400.00'), Decimal('4400.00'), Decimal('4400.00')),
 (202, 20, Decimal('6000.00'), Decimal('19000.00'), Decimal('9500.00')),
 (201, 20, Decimal('13000.00'), Decimal('19000.00'), Decimal('9500.00')),
 (119, 30, Decimal('2500.00'), Decimal('24900.00'), Decimal('4150.00')),
 (118, 30, Decimal('2600.00'), Decimal('24900.00'), Decimal('4150.00')),
 (117, 30, Decimal('2800.00'), Decimal('24900.00'), Decimal('4150.00')),
 (116, 30, Decimal('2900.00'), Decimal('24900.00'), Decimal('4150.00')),
 (115, 30, Decimal('3100.00'), Decimal('24900.00'), Decimal('4150.00')),
 (114, 30, Decimal('11000.00'), Decimal('24900.00'), Decimal('4150.00')),
 (203, 40, Decimal('6500.00'), Decimal('6500.00'), Decimal('6500.00'))]

In [8]:
%%sql

SELECT e.employee_id, e.department_id, e.salary,
    ae.department_salary_expense,
    ae.avg_salary_expense,
    round(e.salary/ae.department_salary_expense * 100, 2) pct_salary
FROM employees e JOIN (
    SELECT department_id, 
        sum(salary) AS department_salary_expense,
        round(avg(salary)::numeric, 2) AS avg_salary_expense
    FROM employees
    GROUP BY department_id
) ae
ON e.department_id = ae.department_id
ORDER BY department_id, salary
LIMIT 10

 * postgresql://itversity_hr_user:***@localhost:5432/itversity_hr_db
10 rows affected.


[(200, 10, Decimal('4400.00'), Decimal('4400.00'), Decimal('4400.00'), Decimal('100.00')),
 (202, 20, Decimal('6000.00'), Decimal('19000.00'), Decimal('9500.00'), Decimal('31.58')),
 (201, 20, Decimal('13000.00'), Decimal('19000.00'), Decimal('9500.00'), Decimal('68.42')),
 (119, 30, Decimal('2500.00'), Decimal('24900.00'), Decimal('4150.00'), Decimal('10.04')),
 (118, 30, Decimal('2600.00'), Decimal('24900.00'), Decimal('4150.00'), Decimal('10.44')),
 (117, 30, Decimal('2800.00'), Decimal('24900.00'), Decimal('4150.00'), Decimal('11.24')),
 (116, 30, Decimal('2900.00'), Decimal('24900.00'), Decimal('4150.00'), Decimal('11.65')),
 (115, 30, Decimal('3100.00'), Decimal('24900.00'), Decimal('4150.00'), Decimal('12.45')),
 (114, 30, Decimal('11000.00'), Decimal('24900.00'), Decimal('4150.00'), Decimal('44.18')),
 (203, 40, Decimal('6500.00'), Decimal('6500.00'), Decimal('6500.00'), Decimal('100.00'))]

```{note}
Let us see how we can get it using Analytics/Windowing Functions. 
```

* We can use all standard aggregate functions such as `count`, `sum`, `min`, `max`, `avg` etc.

In [10]:
%%sql

SELECT e.employee_id, e.department_id, e.salary,
    sum(e.salary) OVER (
        PARTITION BY e.department_id
    ) AS department_salary_expense
FROM employees e
ORDER BY e.department_id
LIMIT 10

 * postgresql://itversity_hr_user:***@localhost:5432/itversity_hr_db
10 rows affected.


[(200, 10, Decimal('4400.00'), Decimal('4400.00')),
 (201, 20, Decimal('13000.00'), Decimal('19000.00')),
 (202, 20, Decimal('6000.00'), Decimal('19000.00')),
 (114, 30, Decimal('11000.00'), Decimal('24900.00')),
 (115, 30, Decimal('3100.00'), Decimal('24900.00')),
 (116, 30, Decimal('2900.00'), Decimal('24900.00')),
 (117, 30, Decimal('2800.00'), Decimal('24900.00')),
 (118, 30, Decimal('2600.00'), Decimal('24900.00')),
 (119, 30, Decimal('2500.00'), Decimal('24900.00')),
 (203, 40, Decimal('6500.00'), Decimal('6500.00'))]

In [12]:
%%sql

SELECT e.employee_id, e.department_id, e.salary,
    sum(e.salary) OVER (
        PARTITION BY e.department_id
    ) AS department_salary_expense,
    round(e.salary / sum(e.salary) OVER (
        PARTITION BY e.department_id
    ) * 100, 2) AS pct_salary
FROM employees e
ORDER BY e.department_id,
    e.salary
LIMIT 10

 * postgresql://itversity_hr_user:***@localhost:5432/itversity_hr_db
10 rows affected.


[(200, 10, Decimal('4400.00'), Decimal('4400.00'), Decimal('100.00')),
 (202, 20, Decimal('6000.00'), Decimal('19000.00'), Decimal('31.58')),
 (201, 20, Decimal('13000.00'), Decimal('19000.00'), Decimal('68.42')),
 (119, 30, Decimal('2500.00'), Decimal('24900.00'), Decimal('10.04')),
 (118, 30, Decimal('2600.00'), Decimal('24900.00'), Decimal('10.44')),
 (117, 30, Decimal('2800.00'), Decimal('24900.00'), Decimal('11.24')),
 (116, 30, Decimal('2900.00'), Decimal('24900.00'), Decimal('11.65')),
 (115, 30, Decimal('3100.00'), Decimal('24900.00'), Decimal('12.45')),
 (114, 30, Decimal('11000.00'), Decimal('24900.00'), Decimal('44.18')),
 (203, 40, Decimal('6500.00'), Decimal('6500.00'), Decimal('100.00'))]

In [17]:
%%sql

SELECT e.employee_id, e.department_id, e.salary,
    sum(e.salary) OVER (
        PARTITION BY e.department_id
    ) AS sum_sal_expense,
    round(avg(e.salary) OVER (
        PARTITION BY e.department_id
    ), 2) AS avg_sal_expense,
    min(e.salary) OVER (
        PARTITION BY e.department_id
    ) AS min_sal_expense,
    max(e.salary) OVER (
        PARTITION BY e.department_id
    ) AS max_sal_expense,
    count(e.salary) OVER (
        PARTITION BY e.department_id
    ) AS cnt_sal_expense
FROM employees e
ORDER BY e.department_id,
    e.salary
LIMIT 10

 * postgresql://itversity_hr_user:***@localhost:5432/itversity_hr_db
10 rows affected.


[(200, 10, Decimal('4400.00'), Decimal('4400.00'), Decimal('4400.00'), Decimal('4400.00'), Decimal('4400.00'), 1),
 (202, 20, Decimal('6000.00'), Decimal('19000.00'), Decimal('9500.00'), Decimal('6000.00'), Decimal('13000.00'), 2),
 (201, 20, Decimal('13000.00'), Decimal('19000.00'), Decimal('9500.00'), Decimal('6000.00'), Decimal('13000.00'), 2),
 (119, 30, Decimal('2500.00'), Decimal('24900.00'), Decimal('4150.00'), Decimal('2500.00'), Decimal('11000.00'), 6),
 (118, 30, Decimal('2600.00'), Decimal('24900.00'), Decimal('4150.00'), Decimal('2500.00'), Decimal('11000.00'), 6),
 (117, 30, Decimal('2800.00'), Decimal('24900.00'), Decimal('4150.00'), Decimal('2500.00'), Decimal('11000.00'), 6),
 (116, 30, Decimal('2900.00'), Decimal('24900.00'), Decimal('4150.00'), Decimal('2500.00'), Decimal('11000.00'), 6),
 (115, 30, Decimal('3100.00'), Decimal('24900.00'), Decimal('4150.00'), Decimal('2500.00'), Decimal('11000.00'), 6),
 (114, 30, Decimal('11000.00'), Decimal('24900.00'), Decimal('4150.00'), Decimal('2500.00'), Decimal('11000.00'), 6),
 (203, 40, Decimal('6500.00'), Decimal('6500.00'), Decimal('6500.00'), Decimal('6500.00'), Decimal('6500.00'), 1)]

```{warning}
If you are using Jupyter based environment make sure to restart the kernel, as the session might have been already connected with hr database.
```

In [3]:
%load_ext sql

In [4]:
%env DATABASE_URL=postgresql://itversity_retail_user:retail_password@localhost:5432/itversity_retail_db

env: DATABASE_URL=postgresql://itversity_retail_user:retail_password@localhost:5432/itversity_retail_db


In [5]:
%%sql

SELECT
    order_date,
    order_item_product_id,
    revenue,
    sum(revenue) OVER (PARTITION BY order_date) AS sum_revenue,
    min(revenue) OVER (PARTITION BY order_date) AS min_revenue,
    max(revenue) OVER (PARTITION BY order_date) AS max_revenue
FROM daily_product_revenue
ORDER BY order_date,
    revenue DESC
LIMIT 10

10 rows affected.


[(datetime.datetime(2013, 7, 25, 0, 0), 1004, Decimal('5599.72'), Decimal('31547.23'), Decimal('49.98'), Decimal('5599.72')),
 (datetime.datetime(2013, 7, 25, 0, 0), 191, Decimal('5099.49'), Decimal('31547.23'), Decimal('49.98'), Decimal('5599.72')),
 (datetime.datetime(2013, 7, 25, 0, 0), 957, Decimal('4499.70'), Decimal('31547.23'), Decimal('49.98'), Decimal('5599.72')),
 (datetime.datetime(2013, 7, 25, 0, 0), 365, Decimal('3359.44'), Decimal('31547.23'), Decimal('49.98'), Decimal('5599.72')),
 (datetime.datetime(2013, 7, 25, 0, 0), 1073, Decimal('2999.85'), Decimal('31547.23'), Decimal('49.98'), Decimal('5599.72')),
 (datetime.datetime(2013, 7, 25, 0, 0), 1014, Decimal('2798.88'), Decimal('31547.23'), Decimal('49.98'), Decimal('5599.72')),
 (datetime.datetime(2013, 7, 25, 0, 0), 403, Decimal('1949.85'), Decimal('31547.23'), Decimal('49.98'), Decimal('5599.72')),
 (datetime.datetime(2013, 7, 25, 0, 0), 502, Decimal('1650.00'), Decimal('31547.23'), Decimal('49.98'), Decimal('5599.72')),
 (datetime.datetime(2013, 7, 25, 0, 0), 627, Decimal('1079.73'), Decimal('31547.23'), Decimal('49.98'), Decimal('5599.72')),
 (datetime.datetime(2013, 7, 25, 0, 0), 226, Decimal('599.99'), Decimal('31547.23'), Decimal('49.98'), Decimal('5599.72'))]

## Cumulative or Moving Aggregations

Let us understand how we can take care of cumulative or moving aggregations using Analytic Functions.
* When it comes to Windowing or Analytic Functions we can also specify window spec using `ROWS BETWEEN` clause.
* Even when we do not specify window spec, the default window spec is used. For most of the functions the default window spec is `UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING`. You also have special clauses such as `CURRENT ROW`.
* Here are some of the examples with respect to `ROWS BETWEEN`.
  * `ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING`
  * `ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW`
  * `ROWS BETWEEN CURRENT ROW AND UNBOUNDED FOLLOWING`
  * `ROWS BETWEEN 3 PRECEDING AND CURRENT ROW` - moving aggregations using current record and previous 3 records.
  * `ROWS BETWEEN CURRENT ROW AND 3 FOLLOWING` - moving aggregations using current record and following 3 records.
  * `ROWS BETWEEN 3 PRECEDING AND 3 FOLLOWING` - moving aggregations based up on 7 records (current record, 3 previous records and 3 following records)
* We can leverage `ROWS BETWEEN` for cumulative aggregations or moving aggregations.
* Here is an example of cumulative sum.

```{warning}
If you are using Jupyter based environment make sure to restart the kernel, as the session might have been already connected with retail database.
```

In [1]:
%load_ext sql

In [2]:
%env DATABASE_URL=postgresql://itversity_hr_user:hr_password@localhost:5432/itversity_hr_db

env: DATABASE_URL=postgresql://itversity_hr_user:hr_password@localhost:5432/itversity_hr_db


```{note}
Even though it is not mandatory to specify `ORDER BY` as per syntax for cumulative aggregations, it is a must to specify. If not, you will end up getting incorrect results.
```

In [4]:
%%sql

SELECT e.employee_id, e.department_id, e.salary,
    sum(e.salary) OVER (
        PARTITION BY e.department_id
        ORDER BY e.salary
        ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW
    ) AS sum_sal_expense
FROM employees e
ORDER BY e.department_id, e.salary DESC
LIMIT 10

 * postgresql://itversity_hr_user:***@localhost:5432/itversity_hr_db
10 rows affected.


[(200, 10, Decimal('4400.00'), Decimal('4400.00')),
 (201, 20, Decimal('13000.00'), Decimal('19000.00')),
 (202, 20, Decimal('6000.00'), Decimal('6000.00')),
 (114, 30, Decimal('11000.00'), Decimal('24900.00')),
 (115, 30, Decimal('3100.00'), Decimal('13900.00')),
 (116, 30, Decimal('2900.00'), Decimal('10800.00')),
 (117, 30, Decimal('2800.00'), Decimal('7900.00')),
 (118, 30, Decimal('2600.00'), Decimal('5100.00')),
 (119, 30, Decimal('2500.00'), Decimal('2500.00')),
 (203, 40, Decimal('6500.00'), Decimal('6500.00'))]

```{warning}
If you are using Jupyter based environment make sure to restart the kernel, as the session might have been already connected with hr database.
```

In [1]:
%load_ext sql

In [2]:
%env DATABASE_URL=postgresql://itversity_retail_user:retail_password@localhost:5432/itversity_retail_db

env: DATABASE_URL=postgresql://itversity_retail_user:retail_password@localhost:5432/itversity_retail_db


```{note}
Here is the example for cumulative sum for every month using daily_product_revenue in retail database.
```

In [3]:
%%sql

SELECT t.*,
    round(sum(t.revenue) OVER (
        PARTITION BY to_char(order_date, 'yyyy-MM')
        ORDER BY order_date
        ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW
    ), 2) AS cumulative_daily_revenue
FROM daily_revenue t
ORDER BY to_char(order_date, 'yyyy-MM'),
    order_date
LIMIT 10

10 rows affected.


[(datetime.datetime(2013, 7, 25, 0, 0), Decimal('31547.23'), Decimal('31547.23')),
 (datetime.datetime(2013, 7, 26, 0, 0), Decimal('54713.23'), Decimal('86260.46')),
 (datetime.datetime(2013, 7, 27, 0, 0), Decimal('48411.48'), Decimal('134671.94')),
 (datetime.datetime(2013, 7, 28, 0, 0), Decimal('35672.03'), Decimal('170343.97')),
 (datetime.datetime(2013, 7, 29, 0, 0), Decimal('54579.70'), Decimal('224923.67')),
 (datetime.datetime(2013, 7, 30, 0, 0), Decimal('49329.29'), Decimal('274252.96')),
 (datetime.datetime(2013, 7, 31, 0, 0), Decimal('59212.49'), Decimal('333465.45')),
 (datetime.datetime(2013, 8, 1, 0, 0), Decimal('49160.08'), Decimal('49160.08')),
 (datetime.datetime(2013, 8, 2, 0, 0), Decimal('50688.58'), Decimal('99848.66')),
 (datetime.datetime(2013, 8, 3, 0, 0), Decimal('43416.74'), Decimal('143265.40'))]

```{note}
Here are examples for 3 day moving sum as well as average using daily_revenue in retail database.
```

In [5]:
%%sql

SELECT t.*,
    round(sum(t.revenue) OVER (
        ORDER BY order_date
        ROWS BETWEEN 2 PRECEDING AND CURRENT ROW
    ), 2) AS moving_3day_revenue
FROM daily_revenue t
ORDER BY order_date
LIMIT 20

 * postgresql://itversity_retail_user:***@localhost:5432/itversity_retail_db
20 rows affected.


[(datetime.datetime(2013, 7, 25, 0, 0), Decimal('31547.23'), Decimal('31547.23')),
 (datetime.datetime(2013, 7, 26, 0, 0), Decimal('54713.23'), Decimal('86260.46')),
 (datetime.datetime(2013, 7, 27, 0, 0), Decimal('48411.48'), Decimal('134671.94')),
 (datetime.datetime(2013, 7, 28, 0, 0), Decimal('35672.03'), Decimal('138796.74')),
 (datetime.datetime(2013, 7, 29, 0, 0), Decimal('54579.70'), Decimal('138663.21')),
 (datetime.datetime(2013, 7, 30, 0, 0), Decimal('49329.29'), Decimal('139581.02')),
 (datetime.datetime(2013, 7, 31, 0, 0), Decimal('59212.49'), Decimal('163121.48')),
 (datetime.datetime(2013, 8, 1, 0, 0), Decimal('49160.08'), Decimal('157701.86')),
 (datetime.datetime(2013, 8, 2, 0, 0), Decimal('50688.58'), Decimal('159061.15')),
 (datetime.datetime(2013, 8, 3, 0, 0), Decimal('43416.74'), Decimal('143265.40')),
 (datetime.datetime(2013, 8, 4, 0, 0), Decimal('35093.01'), Decimal('129198.33')),
 (datetime.datetime(2013, 8, 5, 0, 0), Decimal('34025.27'), Decimal('112535.02')),
 (datetime.datetime(2013, 8, 6, 0, 0), Decimal('57843.89'), Decimal('126962.17')),
 (datetime.datetime(2013, 8, 7, 0, 0), Decimal('45525.59'), Decimal('137394.75')),
 (datetime.datetime(2013, 8, 8, 0, 0), Decimal('33549.47'), Decimal('136918.95')),
 (datetime.datetime(2013, 8, 9, 0, 0), Decimal('29225.16'), Decimal('108300.22')),
 (datetime.datetime(2013, 8, 10, 0, 0), Decimal('46435.04'), Decimal('109209.67')),
 (datetime.datetime(2013, 8, 11, 0, 0), Decimal('31155.50'), Decimal('106815.70')),
 (datetime.datetime(2013, 8, 12, 0, 0), Decimal('59014.74'), Decimal('136605.28')),
 (datetime.datetime(2013, 8, 13, 0, 0), Decimal('17956.88'), Decimal('108127.12'))]

In [7]:
%%sql

SELECT t.*,
    round(sum(t.revenue) OVER (
        ORDER BY order_date
        ROWS BETWEEN 2 PRECEDING AND 2 FOLLOWING
    ), 2) AS moving_3day_revenue
FROM daily_revenue t
ORDER BY order_date
LIMIT 20

 * postgresql://itversity_retail_user:***@localhost:5432/itversity_retail_db
20 rows affected.


[(datetime.datetime(2013, 7, 25, 0, 0), Decimal('31547.23'), Decimal('134671.94')),
 (datetime.datetime(2013, 7, 26, 0, 0), Decimal('54713.23'), Decimal('170343.97')),
 (datetime.datetime(2013, 7, 27, 0, 0), Decimal('48411.48'), Decimal('224923.67')),
 (datetime.datetime(2013, 7, 28, 0, 0), Decimal('35672.03'), Decimal('242705.73')),
 (datetime.datetime(2013, 7, 29, 0, 0), Decimal('54579.70'), Decimal('247204.99')),
 (datetime.datetime(2013, 7, 30, 0, 0), Decimal('49329.29'), Decimal('247953.59')),
 (datetime.datetime(2013, 7, 31, 0, 0), Decimal('59212.49'), Decimal('262970.14')),
 (datetime.datetime(2013, 8, 1, 0, 0), Decimal('49160.08'), Decimal('251807.18')),
 (datetime.datetime(2013, 8, 2, 0, 0), Decimal('50688.58'), Decimal('237570.90')),
 (datetime.datetime(2013, 8, 3, 0, 0), Decimal('43416.74'), Decimal('212383.68')),
 (datetime.datetime(2013, 8, 4, 0, 0), Decimal('35093.01'), Decimal('221067.49')),
 (datetime.datetime(2013, 8, 5, 0, 0), Decimal('34025.27'), Decimal('215904.50')),
 (datetime.datetime(2013, 8, 6, 0, 0), Decimal('57843.89'), Decimal('206037.23')),
 (datetime.datetime(2013, 8, 7, 0, 0), Decimal('45525.59'), Decimal('200169.38')),
 (datetime.datetime(2013, 8, 8, 0, 0), Decimal('33549.47'), Decimal('212579.15')),
 (datetime.datetime(2013, 8, 9, 0, 0), Decimal('29225.16'), Decimal('185890.76')),
 (datetime.datetime(2013, 8, 10, 0, 0), Decimal('46435.04'), Decimal('199379.91')),
 (datetime.datetime(2013, 8, 11, 0, 0), Decimal('31155.50'), Decimal('183787.32')),
 (datetime.datetime(2013, 8, 12, 0, 0), Decimal('59014.74'), Decimal('196605.61')),
 (datetime.datetime(2013, 8, 13, 0, 0), Decimal('17956.88'), Decimal('199737.25'))]

In [8]:
%%sql

SELECT t.*,
    round(avg(t.revenue) OVER (
        ORDER BY order_date
        ROWS BETWEEN 2 PRECEDING AND CURRENT ROW
    ), 2) AS moving_3day_revenue
FROM daily_revenue t
ORDER BY order_date
LIMIT 20

 * postgresql://itversity_retail_user:***@localhost:5432/itversity_retail_db
20 rows affected.


[(datetime.datetime(2013, 7, 25, 0, 0), Decimal('31547.23'), Decimal('31547.23')),
 (datetime.datetime(2013, 7, 26, 0, 0), Decimal('54713.23'), Decimal('43130.23')),
 (datetime.datetime(2013, 7, 27, 0, 0), Decimal('48411.48'), Decimal('44890.65')),
 (datetime.datetime(2013, 7, 28, 0, 0), Decimal('35672.03'), Decimal('46265.58')),
 (datetime.datetime(2013, 7, 29, 0, 0), Decimal('54579.70'), Decimal('46221.07')),
 (datetime.datetime(2013, 7, 30, 0, 0), Decimal('49329.29'), Decimal('46527.01')),
 (datetime.datetime(2013, 7, 31, 0, 0), Decimal('59212.49'), Decimal('54373.83')),
 (datetime.datetime(2013, 8, 1, 0, 0), Decimal('49160.08'), Decimal('52567.29')),
 (datetime.datetime(2013, 8, 2, 0, 0), Decimal('50688.58'), Decimal('53020.38')),
 (datetime.datetime(2013, 8, 3, 0, 0), Decimal('43416.74'), Decimal('47755.13')),
 (datetime.datetime(2013, 8, 4, 0, 0), Decimal('35093.01'), Decimal('43066.11')),
 (datetime.datetime(2013, 8, 5, 0, 0), Decimal('34025.27'), Decimal('37511.67')),
 (datetime.datetime(2013, 8, 6, 0, 0), Decimal('57843.89'), Decimal('42320.72')),
 (datetime.datetime(2013, 8, 7, 0, 0), Decimal('45525.59'), Decimal('45798.25')),
 (datetime.datetime(2013, 8, 8, 0, 0), Decimal('33549.47'), Decimal('45639.65')),
 (datetime.datetime(2013, 8, 9, 0, 0), Decimal('29225.16'), Decimal('36100.07')),
 (datetime.datetime(2013, 8, 10, 0, 0), Decimal('46435.04'), Decimal('36403.22')),
 (datetime.datetime(2013, 8, 11, 0, 0), Decimal('31155.50'), Decimal('35605.23')),
 (datetime.datetime(2013, 8, 12, 0, 0), Decimal('59014.74'), Decimal('45535.09')),
 (datetime.datetime(2013, 8, 13, 0, 0), Decimal('17956.88'), Decimal('36042.37'))]

## Analytic Functions – Windowing

Let us go through the list of Windowing functions supported by Postgres.
* `lead` and `lag`
* `first_value` and `last_value`
* We can either use `ORDER BY sort_column` or `PARTITION BY partition_column ORDER BY sort_column` while using Windowing Functions.

In [ ]:
%load_ext sql

### Getting LEAD and LAG values

Let us understand LEAD and LAG functions to get column values from following or prior records.

Here is the example where we can get prior or following records based on **ORDER BY** within **OVER** Clause.

In [ ]:
%env DATABASE_URL=postgresql://itversity_retail_user:retail_password@localhost:5432/itversity_retail_db

In [ ]:
%%sql

SELECT t.*,
    lead(order_date) OVER (ORDER BY order_date DESC) AS prior_date,
    lead(revenue) OVER (ORDER BY order_date DESC) AS prior_revenue,
    lag(order_date) OVER (ORDER BY order_date) AS lag_prior_date,
    lag(revenue) OVER (ORDER BY order_date) AS lag_prior_revenue
FROM daily_revenue AS t
ORDER BY order_date DESC
LIMIT 10

In [ ]:
%%sql

SELECT t.*,
    lead(order_date, 7) OVER (ORDER BY order_date DESC) AS prior_date,
    lead(revenue, 7) OVER (ORDER BY order_date DESC) AS prior_revenue
FROM daily_revenue t
ORDER BY order_date DESC
LIMIT 10

In [ ]:
%%sql

SELECT t.*,
    lead(order_date, 7) OVER (ORDER BY order_date DESC) AS prior_date,
    lead(revenue, 7) OVER (ORDER BY order_date DESC) AS prior_revenue
FROM daily_revenue t
ORDER BY order_date
LIMIT 10

In [ ]:
%%sql

SELECT t.*,
    lead(order_date, 7) OVER (ORDER BY order_date DESC) AS prior_date,
    lead(revenue, 7, 0.0) OVER (ORDER BY order_date DESC) AS prior_revenue
FROM daily_revenue t
ORDER BY order_date
LIMIT 10

In [ ]:
%%sql

SELECT * FROM daily_product_revenue 
ORDER BY order_date, revenue DESC
LIMIT 10

In [ ]:
%%sql

SELECT t.*,
    LEAD(order_item_product_id) OVER (
        PARTITION BY order_date 
        ORDER BY revenue DESC
    ) next_product_id,
    LEAD(revenue) OVER (
        PARTITION BY order_date 
        ORDER BY revenue DESC
    ) next_revenue
FROM daily_product_revenue t
ORDER BY order_date, revenue DESC
LIMIT 30

### Getting first and last values

Let us see how we can get first and last value based on the criteria. We can also use min or max as well.

Here is the example of using first_value.

In [ ]:
%%sql

SELECT t.*,
    first_value(order_item_product_id) OVER (
        PARTITION BY order_date ORDER BY revenue DESC
    ) first_product_id,
    first_value(revenue) OVER (
        PARTITION BY order_date ORDER BY revenue DESC
    ) first_revenue
FROM daily_product_revenue t
ORDER BY order_date, revenue DESC
LIMIT 10

Let us see an example with last_value. While using last_value we need to specify **ROWS BETWEEN CURRENT ROW AND UNBOUNDED FOLLOWING/PRECEEDING**.
* By default it uses `ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW`.
* The last value with in `UNBOUNDED PRECEDING AND CURRENT ROW` will be current record.
* To get the right value, we have to change the windowing clause to `ROWS BETWEEN CURRENT ROW AND UNBOUNDED FOLLOWING`.

In [ ]:
%%sql

SELECT t.*,
    last_value(order_item_product_id) OVER (
        PARTITION BY order_date ORDER BY revenue
        ROWS BETWEEN CURRENT ROW AND UNBOUNDED FOLLOWING
    ) last_product_id,
    last_value(revenue) OVER (
        PARTITION BY order_date ORDER BY revenue
        ROWS BETWEEN CURRENT ROW AND UNBOUNDED FOLLOWING
    ) last_revenue
FROM daily_product_revenue AS t
ORDER BY order_date, revenue DESC
LIMIT 30

## Analytic Functions – Ranking

Let us see how we can get sparse ranks using **rank** function.

* If we have to get ranks globally, we just need to specify **ORDER BY**
* If we have to get ranks with in a key then we need to specify **PARTITION BY** and then **ORDER BY**.
* By default **ORDER BY** will sort the data in ascending order. We can change the order by passing **DESC** after order by.

Here is an example to assign sparse ranks using daily_product_revenue with in each day based on revenue.

In [ ]:
%load_ext sql

In [ ]:
%env DATABASE_URL=postgresql://itversity_retail_user:retail_password@localhost:5432/itversity_retail_db

In [ ]:
%%sql

SELECT t.*,
    rank() OVER (
        PARTITION BY order_date
        ORDER BY revenue DESC
    ) AS rnk
FROM daily_product_revenue t
ORDER BY order_date, revenue DESC
LIMIT 30

Here is another example to assign sparse ranks using employees data set with in each department.

In [ ]:
%%sql

SELECT
    employee_id,
    department_id,
    salary,
    rank() OVER (
        PARTITION BY department_id
        ORDER BY salary DESC
      ) rnk,
    dense_rank() OVER (
        PARTITION BY department_id
        ORDER BY salary DESC
      ) drnk,
    row_number() OVER (
        PARTITION BY department_id
        ORDER BY salary DESC
      ) rn
FROM employees
ORDER BY department_id, salary DESC

In [ ]:
%%sql

SELECT * FROM employees ORDER BY salary LIMIT 10

In [ ]:
%%sql

SELECT employee_id, salary,
    dense_rank() OVER (ORDER BY salary DESC) AS drnk
FROM employees

Let us understand the difference between **rank**, **dense_rank** and **row_number**.

* We can either of the functions to generate ranks when the rank field does not have duplicates.
* When rank field have duplicates then row_number should not be used as it generate unique number for each record with in the partition.
* **rank** will skip the ranks in between if multiple people get the same rank while **dense_rank** continue with the next number.

In [ ]:
%%sql

SELECT
    t.*,
    rank() OVER (
        PARTITION BY order_date
        ORDER BY revenue DESC
    ) rnk,
    dense_rank() OVER (
        PARTITION BY order_date
        ORDER BY revenue DESC
    ) drnk,
    row_number() OVER (
        PARTITION BY order_date
        ORDER BY revenue DESC
    ) rn
FROM daily_product_revenue AS t
ORDER BY order_date, revenue DESC
LIMIT 30

## Final Solution – Top 5 Daily Products

Let us go through the final solution for getting top 5 daily products based up on the revenue.

In [ ]:
%load_ext sql

In [ ]:
%env DATABASE_URL=postgresql://itversity_retail_user:retail_password@localhost:5432/itversity_retail_db

### Overview of Sub Queries

Let us recap about Sub Queries.

* We typically have Sub Queries in **FROM** Clause.
* We need to provide alias to the Sub Queries in **FROM** Clause in Postgresql.
* We use sub queries quite often over queries using Analytics/Windowing Functions

In [ ]:
%%sql

SELECT * FROM (SELECT current_date) AS q

Let us see few more examples with respected to Nested Sub Queries.

In [ ]:
%%sql

SELECT * FROM (
  SELECT order_date, count(1) AS order_count
  FROM orders
  GROUP BY order_date
) q
ORDER BY order_date
LIMIT 10

In [ ]:
%%sql

SELECT * FROM (
  SELECT order_date, count(1) AS order_count
  FROM orders
  GROUP BY order_date
) q
WHERE q.order_count > 0
ORDER BY order_date
LIMIT 10

```{note}
Above query is an example for sub queries. We can achieve using HAVING clause (no need to have sub query to filter)
```

### Filtering - Analytic Function Results

Let us understand how to filter on top of results of Analytic Functions.

* We can use Analytic Functions only in **SELECT** Clause.
* If we have to filter based on Analytic Function results, then we need to use Sub Queries.
* Once the query is nested, we can apply filter using aliases of the Analytic Functions.

Here is the example where we can filter data based on Analytic Functions.

In [ ]:
%%sql

SELECT * FROM (
  SELECT t.*,
    dense_rank() OVER (
      PARTITION BY order_date
      ORDER BY revenue DESC
    ) AS drnk
  FROM daily_product_revenue t
) q
WHERE drnk <= 5
ORDER BY q.order_date, q.revenue DESC
LIMIT 10

### Ranking and Filtering - Recap

Let us recap the procedure to get top 5 orders by revenue for each day.

* We have our original data in **orders** and **order_items**
* We can pre-compute the data or create a view with the logic to generate **daily product revenue**
* Then, we have to use the view or table or even nested query to compute rank
* Once the ranks are computed, we need to nest it to filter based up on our requirement.
* Let us see using the query example.

Let us come up with the query to compute daily product revenue.

In [ ]:
%%sql

SELECT o.order_date,
       oi.order_item_product_id,
       round(sum(oi.order_item_subtotal)::numeric, 2) AS revenue
FROM orders o JOIN order_items oi
ON o.order_id = oi.order_item_order_id
WHERE o.order_status IN ('COMPLETE', 'CLOSED')
GROUP BY o.order_date, oi.order_item_product_id
ORDER BY o.order_date, revenue DESC
LIMIT 30

Let us compute the rank for each product with in each date using revenue as criteria.

In [ ]:
%%sql


SELECT nq.*,
    dense_rank() OVER (
        PARTITION BY order_date
        ORDER BY revenue DESC
    ) AS drnk
FROM (
    SELECT o.order_date,
        oi.order_item_product_id,
        round(sum(oi.order_item_subtotal)::numeric, 2) AS revenue
    FROM orders o 
        JOIN order_items oi
            ON o.order_id = oi.order_item_order_id
    WHERE o.order_status IN ('COMPLETE', 'CLOSED')
    GROUP BY o.order_date, oi.order_item_product_id
) nq
ORDER BY order_date, revenue DESC
LIMIT 30

Now let us see how we can filter the data.

In [ ]:
%%sql

SELECT * FROM (
    SELECT nq.*,
        dense_rank() OVER (
            PARTITION BY order_date
            ORDER BY revenue DESC
        ) AS drnk
    FROM (
        SELECT o.order_date,
            oi.order_item_product_id,
            round(sum(oi.order_item_subtotal)::numeric, 2) AS revenue
        FROM orders o 
            JOIN order_items oi
                ON o.order_id = oi.order_item_order_id
        WHERE o.order_status IN ('COMPLETE', 'CLOSED')
        GROUP BY o.order_date, oi.order_item_product_id
    ) nq
) nq1
WHERE drnk <= 5
ORDER BY order_date, revenue DESC
LIMIT 20

In [ ]:
%%sql

SELECT * FROM (SELECT dpr.*,
  dense_rank() OVER (
    PARTITION BY order_date
    ORDER BY revenue DESC
  ) AS drnk
FROM daily_product_revenue AS dpr) q
WHERE drnk <= 5
ORDER BY order_date, revenue DESC
LIMIT 20

## Exercises - Analytics Functions

Let us take care of the exercises related to analytics functions. We will be using HR database for the same.

* Get all the employees who is making more than average salary with in each department.
* Get cumulative salary for one of the department along with department name.
* Get top 3 paid employees with in each department by salary (use dense_rank)
* Get top 3 products sold in the month of 2014 January by revenue.
* Get top 3 products in each category sold in the month of 2014 January by revenue.

### Prepare HR Database

Here are the steps to prepare HR database.
* Connect to HR DB using `psql` or SQL Workbench. Here is the sample `psql` command.

```shell
psql -h localhost \
    -p 5432 \
    -d itversity_hr_db \
    -U itversity_hr_user \
    -W
```

* Run scripts to create tables and load the data. You can also drop the tables if they already exists.

```sql
\i /data/hr_db/drop_tables_pg.sql
\i /data/hr_db/create_tables_pg.sql
\i /data/hr_db/load_tables_pg.sql
```

* Validate to ensure that data is available in the tables by running these queries.

In [ ]:
%load_ext sql

In [ ]:
%env DATABASE_URL=postgresql://itversity_hr_user:hr_password@localhost:5432/itversity_hr_db

In [ ]:
%sql SELECT * FROM employees LIMIT 10

In [ ]:
%%sql 

SELECT * FROM departments 
ORDER BY manager_id NULLS LAST
LIMIT 10

### Exercise 1

Get all the employees who is making more than average salary with in each department.

* Use HR database employees and department tables for this problem.
* Compute average salary expense for each department and get those employee details who are making more salary than average salary.
* Make sure average salary expense per department is rounded off to 2 decimals.
* Output should contain employee_id, department_name, salary and avg_salary_expense (derived field).
* Data should be sorted in ascending order by department_id and descending order by salary.

|employee_id|department_name|salary|avg_salary_expense|
|---|---|---|---|
|201|Marketing|13000.00|9500.00|
|114|Purchasing|11000.00|4150.00|
|121|Shipping|8200.00|3475.56|
|120|Shipping|8000.00|3475.56|
|122|Shipping|7900.00|3475.56|
|123|Shipping|6500.00|3475.56|
|124|Shipping|5800.00|3475.56|
|184|Shipping|4200.00|3475.56|
|185|Shipping|4100.00|3475.56|
|192|Shipping|4000.00|3475.56|
|193|Shipping|3900.00|3475.56|
|188|Shipping|3800.00|3475.56|
|137|Shipping|3600.00|3475.56|
|189|Shipping|3600.00|3475.56|
|141|Shipping|3500.00|3475.56|
|103|IT|9000.00|5760.00|
|104|IT|6000.00|5760.00|
|145|Sales|14000.00|8955.88|
|146|Sales|13500.00|8955.88|
|147|Sales|12000.00|8955.88|
|168|Sales|11500.00|8955.88|
|148|Sales|11000.00|8955.88|
|174|Sales|11000.00|8955.88|
|149|Sales|10500.00|8955.88|
|162|Sales|10500.00|8955.88|
|156|Sales|10000.00|8955.88|
|150|Sales|10000.00|8955.88|
|169|Sales|10000.00|8955.88|
|170|Sales|9600.00|8955.88|
|163|Sales|9500.00|8955.88|
|151|Sales|9500.00|8955.88|
|157|Sales|9500.00|8955.88|
|158|Sales|9000.00|8955.88|
|152|Sales|9000.00|8955.88|
|100|Executive|24000.00|19333.33|
|108|Finance|12000.00|8600.00|
|109|Finance|9000.00|8600.00|
|205|Accounting|12000.00|10150.00|

In [ ]:
%load_ext sql

In [ ]:
%env DATABASE_URL=postgresql://itversity_hr_user:hr_password@localhost:5432/itversity_hr_db

### Exercise 2

Get cumulative salary with in each department for Finance and IT department along with department name.

* Use HR database employees and department tables for this problem.
* Compute cumulative salary expense for **Finance** as well as **IT** departments with in respective departments.
* Make sure cumulative salary expense per department is rounded off to 2 decimals.
* Output should contain employee_id, department_name, salary and cum_salary_expense (derived field).
* Data should be sorted in ascending order by department_name and then salary.

|employee_id|department_name|salary|cum_salary_expense|
|---|---|---|---|
|113|Finance|6900.00|6900.00|
|111|Finance|7700.00|14600.00|
|112|Finance|7800.00|22400.00|
|110|Finance|8200.00|30600.00|
|109|Finance|9000.00|39600.00|
|108|Finance|12000.00|51600.00|
|107|IT|4200.00|4200.00|
|106|IT|4800.00|9000.00|
|105|IT|4800.00|13800.00|
|104|IT|6000.00|19800.00|
|103|IT|9000.00|28800.00|

### Exercise 3

Get top 3 paid employees with in each department by salary (use dense_rank)

* Use HR database employees and department tables for this problem.
* Highest paid employee should be ranked first.
* Output should contain employee_id, department_id, department_name, salary and employee_rank (derived field).
* Data should be sorted in ascending order by department_id in ascending order and then salary in descending order.

|employee_id|department_id|department_name|salary|employee_rank|
|---|---|---|---|---|
|200|10|Administration|4400.00|1|
|201|20|Marketing|13000.00|1|
|202|20|Marketing|6000.00|2|
|114|30|Purchasing|11000.00|1|
|115|30|Purchasing|3100.00|2|
|116|30|Purchasing|2900.00|3|
|203|40|Human Resources|6500.00|1|
|121|50|Shipping|8200.00|1|
|120|50|Shipping|8000.00|2|
|122|50|Shipping|7900.00|3|
|103|60|IT|9000.00|1|
|104|60|IT|6000.00|2|
|105|60|IT|4800.00|3|
|106|60|IT|4800.00|3|
|204|70|Public Relations|10000.00|1|
|145|80|Sales|14000.00|1|
|146|80|Sales|13500.00|2|
|147|80|Sales|12000.00|3|
|100|90|Executive|24000.00|1|
|101|90|Executive|17000.00|2|
|102|90|Executive|17000.00|2|
|108|100|Finance|12000.00|1|
|109|100|Finance|9000.00|2|
|110|100|Finance|8200.00|3|
|205|110|Accounting|12000.00|1|
|206|110|Accounting|8300.00|2|

### Exercise 4

Get top 3 products sold in the month of 2014 January by revenue.

* Use retail database tables such as orders, order_items and products.
* Highest revenue generating product should come at top.
* Output should contain product_id, product_name, revenue, product_rank. **revenue** and **product_rank** are derived fields.
* Data should be sorted in descending order by revenue.

|product_id|product_name|revenue|product_rank|
|---|---|---|---|
|1004|Field & Stream Sportsman 16 Gun Fire Safe|250787.46|1|
|365|Perfect Fitness Perfect Rip Deck|151474.75|2|
|957|Diamondback Women's Serene Classic Comfort Bi|148190.12|3|


### Exercise 5

Get top 3 products sold in the month of 2014 January under selected categories by revenue. The categories are **Cardio Equipment** and **Strength Training**.

* Use retail database tables such as orders, order_items, products as well as categories.
* Highest revenue generating product should come at top.
* Output should contain category_id, category_name, product_id, product_name, revenue, product_rank. revenue and product_rank are derived fields.
* Data should be sorted in ascending order by category_id and descending order by revenue.

|category_id|category_name|product_id|product_name|revenue|product_rank|
|---|---|---|---|---|---|
|9|Cardio Equipment|191|Nike Men's Free 5.0+ Running Shoe|132286.77|1|
|9|Cardio Equipment|172|Nike Women's Tempo Shorts|870.00|2|
|10|Strength Training|208|SOLE E35 Elliptical|1999.99|1|
|10|Strength Training|203|GoPro HERO3+ Black Edition Camera|1199.97|2|
|10|Strength Training|216|Yakima DoubleDown Ace Hitch Mount 4-Bike Rack|189.00|3|